packages/block-library/src/navigation/create-data-tree.js

```diff
@@ -0,0 +1,38 @@
+/**
+ * Creates a nested, hierarchical tree representation from unstructured data that
+ * has an inherent relationship defined between individual items.
+ *
+ * For example, by default, each element in the dataset should have an `id` and
+ * `parent` property where the `parent` property indicates a relationship between
+ * the current item and another item with a matching `id` propertys.
+ *
+ * This is useful for building linked lists of data from flat data structures.
+ *
+ * @param {Array} dataset linked data to be rearranged into a hierarchical tree based on relational fields.
+ * @param {string} id the property which uniquey identifies each entry within the array.
+ * @param {*} relation the property which identifies how the current item is related to other items in the data (if at all).
+ */
+function createDataTree( dataset, id = 'id', relation = 'parent' ) {
+	const hashTable = Object.create( null );
+	const dataTree = [];
+
+	dataset.forEach( ( data ) => {
+		hashTable[ data[ id ] ] = {
+			...data,
+			children: [],
+		};
+	} );
+
+	dataset.forEach( ( data ) => {
+		if ( data[ relation ] ) {
+			hashTable[ data[ relation ] ].children.push(
+				hashTable[ data[ id ] ]
+			);
+		} else {
+			dataTree.push( hashTable[ data[ id ] ] );
+		}
+	} );
```

Comment:
You could use `for ... of` loops instead:

```suggestion
	for ( const data of dataset ) {
		hashTable[ data[ id ] ] = {
			...data,
			children: [],
		};
	}

	for ( const data of dataset ) {
		if ( data[ relation ] ) {
			hashTable[ data[ relation ] ].children.push(
				hashTable[ data[ id ] ]
			);
		} else {
			dataTree.push( hashTable[ data[ id ] ] );
		}
	}
```

Reply:
@ZebulanStanphill Thanks for your feedback. Much appreciated.

Out of interest is there a particular reason why you're recommending this change? For example, is `for...of` faster? Is it easier to read? I guess it gets rid of an anon function.

Reply:
In my testing, `for ... of` usually is faster, but since we're transpiling down to ES5, I'm mainly suggesting this because I think it's easier to read.